File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.swift
import Foundation

struct Song {
    let number: Int
    let playCount: Int
}

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var dict = [String: [Song]]()
    var cntDict = [String: Int]()
    
    for (idx, (g, p)) in zip(genres, plays).enumerated() {
        cntDict[g, default: 0] += p
        dict[g, default: []].append(Song(number: idx, playCount: p))
    }
    
    var ret = [Int]()
    
    let sortedCnt = cntDict.sorted { $0.value > $1.value }.map { $0.key }
    
    for genre in sortedCnt {
        let sorted = dict[genre]!.sorted { $0.playCount == $1.playCount ? $0.number < $1.number : $0.playCount > $1.playCount }.prefix(2)
        for song in sorted {
            ret.append(song.number)
        }
    }
    
    return ret
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <bits/stdc++.h>

using namespace std;

struct Song {
    int totalByGenre;
    string genre;
    int cnt;
    int idx;
};

bool cmp(const Song& a, const Song& b) {
    // 속한 노래가 많이 재생된 장르부터
    if(a.totalByGenre != b.totalByGenre) return a.totalByGenre > b.totalByGenre;
    // 장르 내에서 많이 재생된 노래부터
    if(a.cnt != b.cnt) return a.cnt > b.cnt;
    // 고유번호가 낮은 노래부터
    return a.idx < b.idx;
}

vector<Song> songs;

unordered_map<string, int> mp;

vector<int> solution(vector<string> genres, vector<int> plays) {
    int n = genres.size();
    for(int i = 0; i < n; i++) {
        mp[genres[i]] += plays[i];  // 장르별로 카운트만 먼저
    }
    for(int i = 0; i < n; i++) {
        songs.push_back({mp[genres[i]], genres[i], plays[i], i});
    }
    
    sort(songs.begin(), songs.end(), cmp);
    
    vector<int> answer;
    // 장르별로 두 개씩만
    unordered_map<string, int> mp2;
    for(int i = 0; i < n; i++) {
        if(mp2[songs[i].genre] < 2) {
            mp2[songs[i].genre]++;
            answer.push_back(songs[i].idx);
        }
    }
    return answer;
}